src/garage/tf/algos/batch_polopt2.py

```diff
@@ -0,0 +1,281 @@
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            path_baselines = np.append(all_path_baselines[idx], 0)
+            deltas = (path['rewards'] + self.discount * path_baselines[1:] -
+                      path_baselines[:-1])
+            path['advantages'] = np_tensor_utils.discount_cumsum(
+                deltas, self.discount * self.gae_lambda)
+            path['deltas'] = deltas
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = np_tensor_utils.discount_cumsum(
+                path['rewards'], self.discount)
+            returns.append(path['returns'])
+
+        obs = np.concatenate([path['observations'] for path in paths])
+        actions = np.concatenate([path['actions'] for path in paths])
+        rewards = np.concatenate([path['rewards'] for path in paths])
+        returns = np.concatenate(returns)
+        baselines = np.concatenate(baselines)
+
+        agent_infos_path = [path['agent_infos'] for path in paths]
+        agent_infos = dict()
+        for key in self.policy.state_info_keys:
+            agent_infos[key] = np.concatenate(
+                [infos[key] for infos in agent_infos_path])
+
+        env_infos = [path['env_infos'] for path in paths]
+        env_infos = tensor_utils.stack_tensor_dict_list([
+            tensor_utils.pad_tensor_dict(p, max_path_length) for p in env_infos
+        ])
+
+        valids = [np.ones_like(path['returns']) for path in paths]
+        lengths = [v.sum() for v in valids]
+
+        average_discounted_return = (np.mean(
+            [path['returns'][0] for path in paths]))
+
+        undiscounted_returns = [sum(path['rewards']) for path in paths]
+        self.episode_reward_mean.extend(undiscounted_returns)
+
+        samples_data = dict(
+            observations=obs,
+            actions=actions,
+            rewards=rewards,
+            baselines=baselines,
+            returns=returns,
+            lengths=lengths,
+            agent_infos=agent_infos,
+            env_infos=env_infos,
+            paths=paths,
+            average_return=np.mean(undiscounted_returns),
+        )
+
+        tabular.record('Iteration', itr)
+        tabular.record('AverageDiscountedReturn', average_discounted_return)
+        tabular.record('AverageReturn', np.mean(undiscounted_returns))
+        tabular.record('Extras/EpisodeRewardMean',
+                       np.mean(self.episode_reward_mean))
+        tabular.record('NumTrajs', len(paths))
+        tabular.record('StdReturn', np.std(undiscounted_returns))
+        tabular.record('MaxReturn', np.max(undiscounted_returns))
+        tabular.record('MinReturn', np.min(undiscounted_returns))
+
+        return samples_data
+
+    def init_opt(self):
```

Comment:
Can you make the following 3 methods abstract? And also this class an ABC?

Reply:
Will do

Reply:
It inherits `RLAlgorithm` which is already an ABC

Reply:
an ABC can inherit from an ABC